### ISSUE TITLE
Unexpected behaviour of `expects.never` on `stub`-ed method

### ISSUE THREAD

**Issue by BrentWheeldon** [freerange/mocha#490]
I'm working on a test that is passing even when it shouldn't. I've simplified it down to a generic example:

```ruby
setup do
  @mock = Mocha::Mock.new(nil)
  @mock.stubs(:a_method)
end

def test_a_method_never_called
  @mock.expects(:a_method).never

  @mock.a_method
end
```

That test passes, which I've found a bit unexpected. We use the `setup` because there are many test cases and most of them just need a simple stub, but in some we want to assert on how many times the method is called.

If I change my test to:

```ruby
setup do
  @mock = Mocha::Mock.new(nil)
  @mock.stubs(:a_method)
end

def test_a_method_never_called
  @mock.unstub(:a_method)
  @mock.expects(:a_method).never

  @mock.a_method
end
```

It fails as expected... But that feels like a bit of a hack?

I wrote a failing test case for [`MockTest`](https://github.com/freerange/mocha/blob/main/test/unit/mock_test.rb) that demonstrates the issue:

```ruby
  def test_expecting_stubbed_method_never_fails
    mock = build_mock
    mock.stubs(:a_method)
    mock.expects(:a_method).never

    mock.a_method

    assert !mock.__verified__?
  end
```

It's entirely possible I'm missing something here so any advice on alternative approaches would be appreciated! (Also, if this is a bug I'm happy to PR a fix if there's guidance on how this might be fixed!)

**Comment by floehopper** [comment]
Hi @BrentWheeldon. Thanks for your question. I'm sorry for the slow reply - really busy with work right now and about to go on holiday.

I haven't looked at your examples in detail, but at first glance I think your scenario is somewhat similar to #171. Since that issue was reported, I've made some changes to the documentation:

From [here](https://mocha.jamesmead.org/#Expectation_matching___invocation_order):

> Stubs and expectations are basically the same thing. A stub is just an expectation of zero or more invocations. The `Expectation#stubs` method is syntactic sugar to make the intent of the test more explicit.
>
> When a method is invoked on a mock object, the mock object searches through its expectations from newest to oldest to find one that matches the invocation. After the invocation, the matching expectation might stop matching further invocations.

And from [here](https://mocha.jamesmead.org/Mocha/Mock.html):

> However, there are some possible “gotchas” caused by this scheme:
> 
> * if you create an expectation and then a stub for the same method, the stub will always override the expectation and the expectation will never be met.
>
> * if you create a stub and then an expectation for the same method, the expectation will match, and when it stops matching the stub will be used instead, possibly masking test failures.
>
> * if you create different expectations for the same method, they will be invoked in the opposite order than that in which they were specified, rather than the same order.

I hope this at least _explains_ why you're seeing the behaviour you are seeing even if it doesn't suggest a solution. I'll try to find some time as soon as I can to suggest one or more alternative approaches to solve your issue. Sorry I don't have time right now. In the meantime it might be worth a quick look through closed issues to see if I've helped someone in this scenario before.

Otherwise in the long termer the aim is to resolve #173  and work on that is in progress in #448.

**Comment by BrentWheeldon** [comment]
> I'm sorry for the slow reply - really busy with work right now and about to go on holiday.

No worries at all! Enjoy your holiday!

> I haven't looked at your examples in detail, but at first glance I think your scenario is somewhat similar to #171. Since that issue was reported, I've made some changes to the documentation:

Yep - it's totally the same issue. I don't know how I missed those docs! Sorry about that. I'll close this out 👍

**Comment by floehopper** [comment]
Thanks! If you can think of a way to improve the docs to prevent other people having the same problem, I'd be really grateful. One thing I've been thinking about is having some kind of cookbook in the docs with some fuller examples than the API docs allow. Do you think that might have helped in this case?

**Comment by BrentWheeldon** [comment]
I'm not sure, to be honest. I really don't know how I missed the existing docs - it was all right there. I think you can chalk this one up to user error!

**Comment by floehopper** [comment]
A fix was released in v2.7.0 - better late than never?!
